File: examples/siwin_shared.py
#!/usr/bin/env python3

import math
import os
import sys
import time
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
GENERATED_DIR = ROOT / "src" / "figdraw" / "bindings" / "generated"
sys.path.insert(0, str(GENERATED_DIR))

import figdraw as fd


COPIES = int(os.environ.get("FIGDRAW_COPIES", "100"))
RUN_ONCE = os.environ.get("FIGDRAW_RUN_ONCE", "").lower() in ("1", "true", "yes", "on")
NO_SLEEP = os.environ.get("FIGDRAW_NO_SLEEP", "1").lower() in ("1", "true", "yes", "on")
TRACE_SHARED = os.environ.get("FIGDRAW_TRACE_SHARED", "").lower() in (
    "1",
    "true",
    "yes",
    "on",
)


RED_FILL = fd.ColorRGBA(220, 40, 40, 155)
RED_STROKE = fd.ColorRGBA(0, 0, 0, 155)
GREEN_SOLID = fd.ColorRGBA(40, 180, 90, 155)
GREEN_GRAD_START = fd.ColorRGBA(18, 112, 64, 255)
GREEN_GRAD_MID = fd.ColorRGBA(40, 180, 90, 255)
GREEN_GRAD_STOP = fd.ColorRGBA(78, 224, 188, 255)
BLUE_SOLID = fd.ColorRGBA(60, 90, 220, 155)
BLUE_GRAD_START = fd.ColorRGBA(44, 72, 186, 255)
BLUE_GRAD_MID = fd.ColorRGBA(60, 90, 220, 255)
BLUE_GRAD_STOP = fd.ColorRGBA(118, 168, 255, 255)
WHITE_STROKE = fd.ColorRGBA(255, 255, 255, 210)
BLACK_SHADOW = fd.ColorRGBA(0, 0, 0, 155)
BLUE_INNER_SHADOW = fd.ColorRGBA(40, 40, 60, 150)
RED_BORDER = 5.0
BLUE_BORDER = 4.0


def fract(x):
    return x - math.floor(x)


def trace(message):
    if TRACE_SHARED:
        print(f"trace: {message}", flush=True)


def require_ref(value, message):
    if getattr(value, "ref", 0) == 0:
        raise RuntimeError(message)
    return value


def build_render_tree(renders, width, height, frame):
    renders.clear()
    t = frame * 0.02

    background = require_ref(
        fd.new_rectangle_fig(0.0, 0.0, width, height),
        "makeRenderTree: newRectangleFig returned nil",
    )
    background.set_fill_color_rgba(fd.ColorRGBA(255, 255, 255, 155))
    renders.add_root(0, background)

    red_start_x = 60.0
    red_start_y = 60.0
    green_start_x = 320.0
    green_start_y = 120.0
    blue_start_x = 180.0
    blue_start_y = 300.0
    max_w = 260.0
    max_h = 180.0
    max_x = max(0.0, width - (green_start_x + max_w))
    max_y = max(0.0, height - (blue_start_y + max_h))

    loop_copies = 1 if TRACE_SHARED else COPIES
    for i in range(loop_copies):
        if TRACE_SHARED and i == 0:
            trace("loop i=0 start")

        seed_x = math.sin(i * 78.233) * 43758.5453
        seed_y = math.sin((i + 19) * 37.719) * 24634.6345
        base_x = fract(seed_x) * max_x if max_x > 0.0 else 0.0
        base_y = fract(seed_y) * max_y if max_y > 0.0 else 0.0
        jitter_x = math.sin(t + i * 0.15) * 20.0
        jitter_y = math.cos(t * 0.9 + i * 0.2) * 20.0
        offset_x = min(max(base_x + jitter_x, 0.0), max_x)
        offset_y = min(max(base_y + jitter_y, 0.0), max_y)
        size_pulse_w = 0.5 + 0.5 * math.sin(t * 0.8 + i * 0.07)
        size_pulse_h = 0.5 + 0.5 * math.cos(t * 0.65 + i * 0.09)
        red_w = 160.0 + 100.0 * size_pulse_w
        red_h = 110.0 + 70.0 * size_pulse_h
        green_w = 160.0 + 100.0 * size_pulse_h
        green_h = 110.0 + 70.0 * size_pulse_w
        blue_w = 160.0 + 100.0 * (1.0 - size_pulse_w)
        blue_h = 110.0 + 70.0 * (1.0 - size_pulse_h)

        red_fig = require_ref(
            fd.new_rectangle_fig(
                red_start_x + offset_x, red_start_y + offset_y, red_w, red_h
            ),
            "newRectangleFig returned nil for red fig",
        )
        red_fig.set_fill_color_rgba(RED_FILL)
        corner_pulse = 0.5 + 0.5 * math.sin(t * 1.25 + i * 0.11)
        c0 = 4.0 + 26.0 * corner_pulse
        c1 = 6.0 + 22.0 * (1.0 - corner_pulse)
        c2 = 8.0 + 18.0 * (0.5 + 0.5 * math.sin(t * 0.7 + i * 0.05))
        c3 = 10.0 + 16.0 * (0.5 + 0.5 * math.cos(t * 0.8 + i * 0.06))
        if not TRACE_SHARED:
            red_fig.set_corners(fd.CornerRadii(c0, c1, c2, c3))
            red_fig.set_stroke(RED_BORDER, RED_STROKE)
        renders.add_root(0, red_fig)

        green_fig = require_ref(
            fd.new_rectangle_fig(
                green_start_x + offset_x,
                green_start_y + offset_y,
                green_w,
                green_h,
            ),
            "newRectangleFig returned nil for green fig",
        )
        if not TRACE_SHARED and i % 2 == 0:
            axis = fd.FGA_X if (i % 4) < 2 else fd.FGA_DIAG_TLBR
            green_fig.set_fill_linear3(
                GREEN_GRAD_START, GREEN_GRAD_MID, GREEN_GRAD_STOP, axis, 128
            )
        else:
            green_fig.set_fill_color_rgba(GREEN_SOLID)

        green_corner_pulse = 0.5 + 0.5 * math.cos(t * 0.95 + i * 0.08)
        g0 = 6.0 + 22.0 * green_corner_pulse
        g1 = 8.0 + 18.0 * (1.0 - green_corner_pulse)
        g2 = 10.0 + 16.0 * (0.5 + 0.5 * math.cos(t * 0.75 + i * 0.04))
        g3 = 12.0 + 14.0 * (0.5 + 0.5 * math.sin(t * 0.85 + i * 0.05))
        shadow_pulse = 0.5 + 0.5 * math.sin(t * 1.1 + i * 0.05)
        shadow_blur = max(0.0, 6.0 + 18.0 * shadow_pulse)
        shadow_spread = max(0.0, 4.0 + 20.0 * (1.0 - shadow_pulse))
        shadow_x = 6.0 + 10.0 * math.sin(t * 0.9 + i * 0.03)
        shadow_y = 6.0 + 10.0 * math.cos(t * 0.9 + i * 0.03)
        if not TRACE_SHARED:
            green_fig.set_corners(fd.CornerRadii(g0, g1, g2, g3))
            green_fig.clear_shadows()
            green_fig.set_shadow(
                0,
                fd.DROP_SHADOW,
                shadow_blur,
                shadow_spread,
                shadow_x,
                shadow_y,
                BLACK_SHADOW,
            )
        renders.add_root(0, green_fig)

        blue_fig = require_ref(
            fd.new_rectangle_fig(
                blue_start_x + offset_x, blue_start_y + offset_y, blue_w, blue_h
            ),
            "newRectangleFig returned nil for blue fig",
        )
        if not TRACE_SHARED and i % 3 == 0:
            axis = fd.FGA_Y if (i % 2) == 0 else fd.FGA_DIAG_BLTR
            blue_fig.set_fill_linear3(
                BLUE_GRAD_START, BLUE_GRAD_MID, BLUE_GRAD_STOP, axis, 132
            )
        else:
            blue_fig.set_fill_color_rgba(BLUE_SOLID)

        inset_pulse = 0.5 + 0.5 * math.sin(t * 1.05 + i * 0.06)
        inset_blur = max(0.0, 8.0 + 10.0 * inset_pulse)
        inset_spread = max(0.0, 2.0 + 10.0 * (1.0 - inset_pulse))
        inset_x = 6.0 * math.sin(t * 0.85 + i * 0.04)
        inset_y = 6.0 * math.cos(t * 0.8 + i * 0.04)
        if not TRACE_SHARED:
            blue_fig.set_stroke(BLUE_BORDER, WHITE_STROKE)
            blue_fig.clear_shadows()
            blue_fig.set_shadow(
                0,
                fd.INNER_SHADOW,
                inset_blur,
                inset_spread,
                inset_x,
                inset_y,
                BLUE_INNER_SHADOW,
            )
        renders.add_root(0, blue_fig)


def main():
    fd.set_fig_data_dir(str(ROOT / "data"))

    typeface = require_ref(
        fd.load_typeface_binding("Ubuntu.ttf"), "Failed to load typeface: Ubuntu.ttf"
    )
    fps_font = require_ref(fd.FigFontRef(typeface, 18.0), "Failed to create fps font")
    fps_text = "0.0 FPS"

    app = require_ref(
        fd.FigSiwinAppRef(
            800,
            600,
            "Siwin RenderList (Python Shared Lib)",
            512,
            1.0,
            False,
            True,
            0,
            True,
            False,
            False,
        ),
        "Failed to create siwin app",
    )
    trace(
        "created siwin app backend="
        + app.siwin_backend_name()
        + " display="
        + app.siwin_display_server_name()
    )
    app.siwin_first_step()

    renders = require_ref(fd.Renders(), "Failed to create renders")

    app_running = True
    global_frame = 0
    frames = 0
    fps_frames = 0
    fps_start = time.time()
    make_render_tree_us_sum = 0.0
    render_frame_us_sum = 0.0
    last_element_count = 0

    try:
        while app.siwin_opened() and app_running:
            app.siwin_refresh_ui_scale()
            frames += 1
            global_frame += 1
            fps_frames += 1

            size = app.siwin_window_size()
            width = float(size.w)
            height = float(size.h)

            t0 = time.perf_counter()
            build_render_tree(renders, width, height, global_frame)
            make_render_tree_us_sum += (time.perf_counter() - t0) * 1_000_000.0
            last_element_count = renders.layer_node_count(0)

            hud_margin = 12.0
            hud_w = 180.0
            hud_h = 34.0
            hud_x = width - hud_w - hud_margin
            hud_y = hud_margin
            hud_rect = require_ref(
                fd.new_rectangle_fig(hud_x, hud_y, hud_w, hud_h),
                "Failed to create HUD rect",
            )
            hud_rect.set_fill_color_rgba(fd.ColorRGBA(0, 0, 0, 155))
            hud_rect.set_corners(fd.CornerRadii(8, 8, 8, 8))
            renders.add_root(0, hud_rect)

            hud_text_pad_x = 10.0
            hud_text_pad_y = 6.0
            hud_text_x = hud_x + hud_text_pad_x
            hud_text_y = hud_y + hud_text_pad_y
            hud_text_w = hud_w - hud_text_pad_x * 2.0
            hud_text_h = hud_h - hud_text_pad_y * 2.0
            fps_layout = fd.typeset_text_binding(
                hud_text_w,
                hud_text_h,
                fps_font,
                fps_text,
                2,
                1,
                False,
                False,
            )
            if getattr(fps_layout, "ref", 0) != 0:
                hud_text = require_ref(
                    fd.new_text_fig(hud_text_x, hud_text_y, hud_text_w, hud_text_h),
                    "Failed to create HUD text",
                )
                hud_text.set_fill_color_rgba(fd.ColorRGBA(0, 0, 0, 0))
                fd.set_fig_text_layout_binding(hud_text, fps_layout)
                renders.add_root(0, hud_text)

            t1 = time.perf_counter()
            app.render_siwin_frame_binding(renders, width, height)
            render_frame_us_sum += (time.perf_counter() - t1) * 1_000_000.0
            app.siwin_redraw()
            app.siwin_step()

            now = time.time()
            elapsed = now - fps_start
            if elapsed >= 1.0:
                fps = fps_frames / elapsed
                fps_text = f"{fps:.1f} FPS"
                avg_make = make_render_tree_us_sum / max(1, fps_frames)
                avg_render = render_frame_us_sum / max(1, fps_frames)
                print(
                    "fps: "
                    f"{fps} | elems: {last_element_count}"
                    f" | makeRenderTree avg(us): {avg_make}"
                    f" | renderFrame avg(us): {avg_render}",
                    flush=True,
                )
                fps_frames = 0
                fps_start = now
                make_render_tree_us_sum = 0.0
                render_frame_us_sum = 0.0

            if RUN_ONCE and frames >= 1:
                app_running = False

            if not NO_SLEEP and app_running:
                time.sleep(0.016)
    finally:
        if not TRACE_SHARED:
            app.siwin_close()


if __name__ == "__main__":
    main()

File: examples/siwin_shared.nim
when defined(emscripten):
  import std/[times, monotimes, strformat]
else:
  import std/[os, times, monotimes, strformat]
import std/math

import figdraw/bindings/generated/figdraw

when defined(macosx):
  {.passL: "-Wl,-rpath,@executable_path/../src/figdraw/bindings/generated".}
  {.passL: "-Wl,-rpath,@loader_path/../src/figdraw/bindings/generated".}
elif defined(linux) or defined(bsd):
  {.passL: "-Wl,-rpath,$ORIGIN/../src/figdraw/bindings/generated".}

const
  Copies = 100
  RunOnce {.booldefine: "figdraw.runOnce".}: bool = false
  NoSleep {.booldefine: "figdraw.noSleep".}: bool = true
  TraceShared {.booldefine: "figdraw.traceSharedLib".}: bool = false

let
  redFill = colorRgba(220, 40, 40, 155)
  redStroke = colorRgba(0, 0, 0, 155)
  greenSolid = colorRgba(40, 180, 90, 155)
  greenGradStart = colorRgba(18, 112, 64, 255)
  greenGradMid = colorRgba(40, 180, 90, 255)
  greenGradStop = colorRgba(78, 224, 188, 255)
  blueSolid = colorRgba(60, 90, 220, 155)
  blueGradStart = colorRgba(44, 72, 186, 255)
  blueGradMid = colorRgba(60, 90, 220, 255)
  blueGradStop = colorRgba(118, 168, 255, 255)
  whiteStroke = colorRgba(255, 255, 255, 210)
  blackShadow = colorRgba(0, 0, 0, 155)
  blueInnerShadow = colorRgba(40, 40, 60, 150)
  redBorder = 5.0'f32
  blueBorder = 4.0'f32

proc fract(x: float64): float64 {.inline.} =
  x - floor(x)

proc buildRenderTree(renders: Renders, w, h: float32, frame: int) =
  if renders.isNil:
    quit("buildRenderTree: renders is nil", 1)
  renders.clear()
  let t = frame.float32 * 0.02'f32

  let background = newRectangleFig(0, 0, w, h)
  GC_ref(background)
  if background.isNil:
    quit("makeRenderTree: newRectangleFig returned nil", 1)
  when TraceShared:
    echo "trace: background created"
  background.setFillColorRgba(colorRgba(255, 255, 255, 155))
  when TraceShared:
    echo "trace: background fill set"
  discard renders.addRoot(0, background)
  when TraceShared:
    echo "trace: background root added"

  let redStartX = 60.0'f32
  let redStartY = 60.0'f32
  let greenStartX = 320.0'f32
  let greenStartY = 120.0'f32
  let blueStartX = 180.0'f32
  let blueStartY = 300.0'f32
  let maxW = 260.0'f32
  let maxH = 180.0'f32
  let maxX = max(0.0'f32, w - (greenStartX + maxW))
  let maxY = max(0.0'f32, h - (blueStartY + maxH))

  let loopCopies = when TraceShared: 1 else: Copies
  for i in 0 ..< loopCopies:
    when TraceShared:
      if i == 0:
        echo "trace: loop i=0 start"
    let
      seedX = sin(i.float64 * 78.233'f64) * 43758.5453'f64
      seedY = sin((i + 19).float64 * 37.719'f64) * 24634.6345'f64
      baseX = (if maxX > 0: (fract(seedX).float32 * maxX) else: 0'f32)
      baseY = (if maxY > 0: (fract(seedY).float32 * maxY) else: 0'f32)
      jitterX = sin((t + i.float32 * 0.15'f32).float64).float32 * 20
      jitterY = cos((t * 0.9'f32 + i.float32 * 0.2'f32).float64).float32 * 20
      offsetX = min(max(baseX + jitterX, 0'f32), maxX)
      offsetY = min(max(baseY + jitterY, 0'f32), maxY)
      sizePulseW =
        0.5'f32 + 0.5'f32 * sin((t * 0.8'f32 + i.float32 * 0.07'f32).float64).float32
      sizePulseH =
        0.5'f32 + 0.5'f32 * cos((t * 0.65'f32 + i.float32 * 0.09'f32).float64).float32
      redW = 160.0'f32 + 100.0'f32 * sizePulseW
      redH = 110.0'f32 + 70.0'f32 * sizePulseH
      greenW = 160.0'f32 + 100.0'f32 * sizePulseH
      greenH = 110.0'f32 + 70.0'f32 * sizePulseW
      blueW = 160.0'f32 + 100.0'f32 * (1.0'f32 - sizePulseW)
      blueH = 110.0'f32 + 70.0'f32 * (1.0'f32 - sizePulseH)

    let redFig = newRectangleFig(redStartX + offsetX, redStartY + offsetY, redW, redH)
    GC_ref(redFig)
    when TraceShared:
      if i == 0:
        echo "trace: red created"
    redFig.setFillColorRgba(redFill)
    let
      cornerPulse =
        0.5'f32 + 0.5'f32 * sin((t * 1.25'f32 + i.float32 * 0.11'f32).float64).float32
      c0 = 4.0'f32 + 26.0'f32 * cornerPulse
      c1 = 6.0'f32 + 22.0'f32 * (1.0'f32 - cornerPulse)
      c2 =
        8.0'f32 +
        18.0'f32 *
        (0.5'f32 + 0.5'f32 * sin((t * 0.7'f32 + i.float32 * 0.05'f32).float64).float32)
      c3 =
        10.0'f32 +
        16.0'f32 *
        (0.5'f32 + 0.5'f32 * cos((t * 0.8'f32 + i.float32 * 0.06'f32).float64).float32)
    when not TraceShared:
      redFig.setCorners(cornerRadii(c0, c1, c2, c3))
      redFig.setStroke(redBorder, redStroke)
    when TraceShared:
      if i == 0:
        echo "trace: red style (trace mode simplified)"
    discard renders.addRoot(0, redFig)
    when TraceShared:
      if i == 0:
        echo "trace: red added"

    let greenFig =
      newRectangleFig(greenStartX + offsetX, greenStartY + offsetY, greenW, greenH)
    GC_ref(greenFig)
    when TraceShared:
      if i == 0:
        echo "trace: green created"
    let useGreenGradient = (i mod 2) == 0
    when not TraceShared:
      if useGreenGradient:
        let axis = (if (i mod 4) < 2: fgaX else: fgaDiagTLBR)
        greenFig.setFillLinear3(
          greenGradStart, greenGradMid, greenGradStop, axis, 128'u8
        )
      else:
        greenFig.setFillColorRgba(greenSolid)
    else:
      greenFig.setFillColorRgba(greenSolid)
    when TraceShared:
      if i == 0:
        echo "trace: green fill"
    let
      greenCornerPulse =
        0.5'f32 + 0.5'f32 * cos((t * 0.95'f32 + i.float32 * 0.08'f32).float64).float32
      g0 = 6.0'f32 + 22.0'f32 * greenCornerPulse
      g1 = 8.0'f32 + 18.0'f32 * (1.0'f32 - greenCornerPulse)
      g2 =
        10.0'f32 +
        16.0'f32 *
        (0.5'f32 + 0.5'f32 * cos((t * 0.75'f32 + i.float32 * 0.04'f32).float64).float32)
      g3 =
        12.0'f32 +
        14.0'f32 *
        (0.5'f32 + 0.5'f32 * sin((t * 0.85'f32 + i.float32 * 0.05'f32).float64).float32)
      shadowPulse =
        0.5'f32 + 0.5'f32 * sin((t * 1.1'f32 + i.float32 * 0.05'f32).float64).float32
      shadowBlur = max(0.0'f32, 6.0'f32 + 18.0'f32 * shadowPulse)
      shadowSpread = max(0.0'f32, 4.0'f32 + 20.0'f32 * (1.0'f32 - shadowPulse))
      shadowX =
        6.0'f32 + 10.0'f32 * sin((t * 0.9'f32 + i.float32 * 0.03'f32).float64).float32
      shadowY =
        6.0'f32 + 10.0'f32 * cos((t * 0.9'f32 + i.float32 * 0.03'f32).float64).float32
    when not TraceShared:
      greenFig.setCorners(cornerRadii(g0, g1, g2, g3))
    when TraceShared:
      if i == 0:
        echo "trace: green corners"
    when not TraceShared:
      greenFig.clearShadows()
    when TraceShared:
      if i == 0:
        echo "trace: green clear shadows"
    when not TraceShared:
      greenFig.setShadow(
        0, DropShadow, shadowBlur, shadowSpread, shadowX, shadowY, blackShadow
      )
    when TraceShared:
      if i == 0:
        echo "trace: green shadow"
    discard renders.addRoot(0, greenFig)
    when TraceShared:
      if i == 0:
        echo "trace: green added"

    let blueFig =
      newRectangleFig(blueStartX + offsetX, blueStartY + offsetY, blueW, blueH)
    GC_ref(blueFig)
    when TraceShared:
      if i == 0:
        echo "trace: blue created"
    let useBlueGradient = (i mod 3) == 0
    when not TraceShared:
      if useBlueGradient:
        let axis = (if (i mod 2) == 0: fgaY else: fgaDiagBLTR)
        blueFig.setFillLinear3(blueGradStart, blueGradMid, blueGradStop, axis, 132'u8)
      else:
        blueFig.setFillColorRgba(blueSolid)
    else:
      blueFig.setFillColorRgba(blueSolid)
    when TraceShared:
      if i == 0:
        echo "trace: blue fill"
    when not TraceShared:
      blueFig.setStroke(blueBorder, whiteStroke)
    when TraceShared:
      if i == 0:
        echo "trace: blue stroke"
    let
      insetPulse =
        0.5'f32 + 0.5'f32 * sin((t * 1.05'f32 + i.float32 * 0.06'f32).float64).float32
      insetBlur = max(0.0'f32, 8.0'f32 + 10.0'f32 * insetPulse)
      insetSpread = max(0.0'f32, 2.0'f32 + 10.0'f32 * (1.0'f32 - insetPulse))
      insetX = 6.0'f32 * sin((t * 0.85'f32 + i.float32 * 0.04'f32).float64).float32
      insetY = 6.0'f32 * cos((t * 0.8'f32 + i.float32 * 0.04'f32).float64).float32
    when not TraceShared:
      blueFig.clearShadows()
    when TraceShared:
      if i == 0:
        echo "trace: blue clear shadows"
    when not TraceShared:
      blueFig.setShadow(
        0, InnerShadow, insetBlur, insetSpread, insetX, insetY, blueInnerShadow
      )
    when TraceShared:
      if i == 0:
        echo "trace: blue shadow"
    discard renders.addRoot(0, blueFig)
    when TraceShared:
      if i == 0:
        echo "trace: blue added"

when isMainModule:
  when not defined(emscripten):
    let libDir = getCurrentDir() / "src" / "figdraw" / "bindings" / "generated"

    proc ensureLib(expected, actual: string) =
      if fileExists(expected):
        return
      if not fileExists(actual):
        return
      try:
        removeFile(expected)
      except OSError:
        discard
      try:
        createSymlink(actual.lastPathPart(), expected)
      except OSError:
        discard

    when defined(macosx):
      let expected = libDir / "libfigdraw.dylib"
      ensureLib(expected, libDir / "libfig_draw.dylib")
      ensureLib(expected, libDir / "libfigdraw.dylib")
    elif defined(linux) or defined(bsd):
      let expected = libDir / "libfigdraw.so"
      ensureLib(expected, libDir / "libfig_draw.so")
      ensureLib(expected, libDir / "libfigdraw.so")
    setFigDataDir(getCurrentDir() / "data")

  var appRunning = true
  var globalFrame = 0

  let typeface = loadTypefaceBinding("Ubuntu.ttf")
  if typeface.isNil:
    quit("Failed to load typeface: Ubuntu.ttf", 1)
  let fpsFont = newFigFontBinding(typeface, 18.0'f32)
  if fpsFont.isNil:
    quit("Failed to create fps font", 1)
  var fpsText = "0.0 FPS"

  let title = "Siwin RenderList (Nim Shared Lib)"
  let app = newFigSiwinAppBinding(
    800'i32, 600'i32, title, 512, 1.0'f32, false, true, 0'i32, true, false, false
  )
  if app.isNil:
    quit("Failed to create siwin app", 1)
  when TraceShared:
    echo "trace: created siwin app backend=",
      app.siwinBackendName(), " display=", app.siwinDisplayServerName()
  app.siwinFirstStep()
  when TraceShared:
    echo "trace: first step"

  var frames = 0
  var fpsFrames = 0
  var fpsStart = epochTime()
  var makeRenderTreeMsSum = 0.0
  var renderFrameMsSum = 0.0
  var lastElementCount = 0
  let renders = newRenders()
  if renders.isNil:
    quit("Failed to create renders", 1)

  try:
    while app.siwinOpened() and appRunning:
      app.siwinRefreshUiScale()
      when TraceShared:
        echo "trace: refresh ui scale"

      inc frames
      inc globalFrame
      inc fpsFrames

      let size = app.siwinWindowSize()
      let width = size.w.float32
      let height = size.h.float32
      when TraceShared:
        echo "trace: logical size ", width, "x", height

      let t0 = getMonoTime()
      buildRenderTree(renders, width, height, globalFrame)
      when TraceShared:
        echo "trace: make render tree done"
      makeRenderTreeMsSum += float((getMonoTime() - t0).inMicroseconds)
      lastElementCount = renders.layerNodeCount(0)

      let
        hudMargin = 12.0'f32
        hudW = 180.0'f32
        hudH = 34.0'f32
        hudX = width - hudW - hudMargin
        hudY = hudMargin
      let hudRect = newRectangleFig(hudX, hudY, hudW, hudH)
      GC_ref(hudRect)
      hudRect.setFillColorRgba(colorRgba(0, 0, 0, 155))
      hudRect.setCorners(cornerRadii(8, 8, 8, 8))
      discard renders.addRoot(0, hudRect)

      let
        hudTextPadX = 10.0'f32
        hudTextPadY = 6.0'f32
        hudTextX = hudX + hudTextPadX
        hudTextY = hudY + hudTextPadY
        hudTextW = hudW - hudTextPadX * 2
        hudTextH = hudH - hudTextPadY * 2

      let fpsLayout = typesetTextBinding(
        hudTextW,
        hudTextH,
        fpsFont,
        fpsText,
        hAlign = 2, # Right
        vAlign = 1, # Middle
        minContent = false,
        wrap = false,
      )
      if not fpsLayout.isNil:
        let hudText = newTextFig(hudTextX, hudTextY, hudTextW, hudTextH)
        GC_ref(hudText)
        hudText.setFillColorRgba(colorRgba(0, 0, 0, 0))
        setFigTextLayoutBinding(hudText, fpsLayout)
        discard renders.addRoot(0, hudText)
      when TraceShared:
        echo "trace: hud text done"

      let t1 = getMonoTime()
      app.renderSiwinFrameBinding(renders, width, height)
      when TraceShared:
        echo "trace: render frame done"
      renderFrameMsSum += float((getMonoTime() - t1).inMicroseconds)
      app.siwinRedraw()
      app.siwinStep()
      when TraceShared:
        echo "trace: step"

      let now = epochTime()
      let elapsed = now - fpsStart
      if elapsed >= 1.0:
        let fps = fpsFrames.float / elapsed
        fpsText = fmt"{fps:0.1f} FPS"
        let avgMake = makeRenderTreeMsSum / max(1, fpsFrames).float
        let avgRender = renderFrameMsSum / max(1, fpsFrames).float
        echo "fps: ",
          fps, " | elems: ", lastElementCount, " | makeRenderTree avg(us): ", avgMake,
          " | renderFrame avg(us): ", avgRender
        fpsFrames = 0
        fpsStart = now
        makeRenderTreeMsSum = 0.0
        renderFrameMsSum = 0.0

      when RunOnce:
        if frames >= 1:
          appRunning = false

      when not NoSleep and not defined(emscripten):
        if appRunning:
          sleep(16)
  finally:
    when not TraceShared:
      app.siwinClose()
